Fix #8817 #8820 Improve CookieRequestCache

majian159 opened the pull request:
fixes #8817 #8820

zeeshanadnan commented:
@majian159 thanks for the fix. I will look into it as well. The build is failing due to the tests.  You have to update the tests as well of `CookieRequestCacheTest`. @eleftherias  @majian159  let me know if i can be of any help.

majian159 commented:
> @majian159 thanks for the fix. I will look into it as well. The build is failing due to the tests. You have to update the tests as well of `CookieRequestCacheTest`. @eleftherias @majian159 let me know if i can be of any help.

Thanks ,i updated the unit test, if you can please help me review.

majian159 commented:
> Thanks for the PR @majian159 and thanks for the help @zeeshanadnan!
> 
> @majian159 Please also add tests for the bugs reported. This will ensure they do not show up again.

I added unit tests, please check.

zeeshanadnan commented:
> > @majian159 thanks for the fix. I will look into it as well. The build is failing due to the tests. You have to update the tests as well of `CookieRequestCacheTest`. @eleftherias @majian159 let me know if i can be of any help.
> 
> Thanks, i updated the unit test, if you can please help me review.

Thanks @majian159 for taking the time and responsibility to fix this. Adding an extra test for cache in not removed by mistake for non matching request as reported would be a good idea. What do you think?

majian159 commented:
> > > @majian159 thanks for the fix. I will look into it as well. The build is failing due to the tests. You have to update the tests as well of `CookieRequestCacheTest`. @eleftherias @majian159 let me know if i can be of any help.
> > 
> > 
> > Thanks, i updated the unit test, if you can please help me review.
> 
> Thanks @majian159 for taking the time and responsibility to fix this. Adding an extra test for cache in not removed by mistake for non matching request would be a good idea. What do you think?

I added the unit test'notMatchingRequestWhenRequestNotContainsSavedRequestCookie' please check.

majian159 commented:
> > > @majian159 thanks for the fix. I will look into it as well. The build is failing due to the tests. You have to update the tests as well of `CookieRequestCacheTest`. @eleftherias @majian159 let me know if i can be of any help.
> > 
> > 
> > Thanks, i updated the unit test, if you can please help me review.
> 
> Thanks @majian159 for taking the time and responsibility to fix this. Adding an extra test for cache in not removed by mistake for non matching request would be a good idea. What do you think?

I added the unit test'notMatchingRequestWhenRequestNotContainsSavedRequestCookie' please check.

majian159 commented:
> Thanks for the PR @majian159 and thanks for the help @zeeshanadnan!
> 
> @majian159 Please also add tests for the bugs reported. This will ensure they do not show up again.

I perfected unit tests, please check.

@eleftherias

eleftherias commented:
@majian159 Please rebase against the master branch.
There are some errors unrelated to your changes that are causing the build to fail.

majian159 commented:
complete

majian159 commented:
> @majian159 Please rebase against the master branch.
> 
> There are some errors unrelated to your changes that are causing the build to fail.

complete.

eleftherias commented on web/src/main/java/org/springframework/security/web/savedrequest/CookieRequestCache.java:
```diff
@@ -115,6 +136,51 @@ private static String decodeCookie(String encodedCookieValue) {
+	private boolean matchesSavedRequest(HttpServletRequest request, SavedRequest savedRequest) {
+		if (savedRequest == null) {
+			return false;
+		} else if (savedRequest instanceof DefaultSavedRequest) {
```
This `else if` block can be removed. We only have information about the saved request URL, so that is all we can really compare.

majian159 replied:
I fixed these issues, please check.

eleftherias commented on web/src/test/java/org/springframework/security/web/savedrequest/CookieRequestCacheTests.java:
```diff
@@ -153,6 +171,7 @@ private MockHttpServletRequest requestToSave() {
 		request.setScheme("https");
 		request.setServerName("abc.com");
 		request.setRequestURI("/destination");
+		request.setServletPath("/destination");
```
We should not have to set the servlet path for this test to pass.

majian159 replied:
I fixed these issues, please check.

eleftherias commented on web/src/test/java/org/springframework/security/web/savedrequest/RequestCacheAwareFilterTests.java:
```diff
@@ -51,6 +51,7 @@ public void doFilterWhenCookieRequestCacheConfiguredThenExpiredSavedRequestCooki
 
 		MockHttpServletRequest request = new MockHttpServletRequest();
 		request.setServerName("abc.com");
+		request.setServletPath("/destination");
```
We should not have to set the servlet path for this test to pass.

majian159 replied:
I fixed these issues, please check.

majian159 commented on web/src/main/java/org/springframework/security/web/savedrequest/CookieRequestCache.java:
```diff
@@ -77,32 +83,47 @@ public SavedRequest getRequest(HttpServletRequest request, HttpServletResponse r
+					.setMethod(request.getMethod())
+					.setRequestURL(getRequestURL(originalURI, uriComponents))
+					.setContextPath(request.getContextPath())
+					.setServletPath(getServletPath(uriComponents, request.getContextPath()))
```
I fixed these issues, please check.

majian159 commented:
I fixed these issues, please check.
@eleftherias

eleftherias commented:
Thanks @majian159! This is now merged into master via 41f26b768a205b29a8c5168e2e1d34dffc887ab4 and fb936e27806dfeb6284359e3cd9c99a2c7a12834.
